drivers/gpio/gpio_pcf857x.c

```diff
@@ -1,69 +1,94 @@
 /**
- * Copyright (c) 2022 Ithinx GmbH
+ * Copyright (c)
+ * 2022 Ithinx GmbH
+ * 2023 Amrith Venkat Kesavamoorthi <amrith@mr-beam.org>
+ * 2023 Mr Beam Lasers GmbH.
  *
  * SPDX-License-Identifier: Apache-2.0
+ *
+ * @see https://www.nxp.com/docs/en/data-sheet/PCF8575.pdf
+ * @see https://www.nxp.com/docs/en/data-sheet/PCF8574_PCF8574A.pdf
  */
 
-#define DT_DRV_COMPAT nxp_pcf8574
+#define DT_DRV_COMPAT nxp_pcf857x
 
 #include <zephyr/drivers/gpio/gpio_utils.h>
 
 #include <zephyr/drivers/gpio.h>
 #include <zephyr/drivers/i2c.h>
 #include <zephyr/kernel.h>
 #include <zephyr/logging/log.h>
-LOG_MODULE_REGISTER(pcf8574, CONFIG_GPIO_LOG_LEVEL);
+LOG_MODULE_REGISTER(pcf857x, CONFIG_GPIO_LOG_LEVEL);
 
-struct pcf8574_pins_cfg {
-	uint8_t configured_as_outputs; /* 0 for input, 1 for output */
-	uint8_t outputs_state;
+struct pcf857x_pins_cfg {
+	uint16_t configured_as_outputs; /* 0 for input, 1 for output */
+	uint16_t outputs_state;
 };
 
-/** Runtime driver data of the pcf8574*/
-struct pcf8574_drv_data {
+/** Runtime driver data of the pcf857x*/
+struct pcf857x_drv_data {
 	/* gpio_driver_data needs to be first */
 	struct gpio_driver_data common;
-	struct pcf8574_pins_cfg pins_cfg;
+	struct pcf857x_pins_cfg pins_cfg;
 	sys_slist_t callbacks;
 	struct k_sem lock;
 	struct k_work work;
 	const struct device *dev;
 	struct gpio_callback int_gpio_cb;
-	uint8_t input_port_last;
+	uint16_t input_port_last;
 };
 
 /** Configuration data*/
-struct pcf8574_drv_cfg {
+struct pcf857x_drv_cfg {
 	/* gpio_driver_config needs to be first */
 	struct gpio_driver_config common;
 	struct i2c_dt_spec i2c;
 	struct gpio_dt_spec gpio_int;
 };
 
 /**
- * @brief Reads the value of the pins from pcf8574 respectively from a connected device.
+ * @brief Reads the value of the pins from pcf857x respectively from a connected device.
  *
  * @param dev Pointer to the device structure of the driver instance.
- * @param value Pointer to the input value. It contains the received Byte.
+ * @param value Pointer to the input value. It contains the received Bytes(receives 2 Bytes for P0
+ * and P1).
  *
  * @retval 0 If successful.
  * @retval Negative value for error code.
  */
-static int pcf8574_process_input(const struct device *dev, gpio_port_value_t *value)
+static int pcf857x_process_input(const struct device *dev, gpio_port_value_t *value)
 {
-	const struct pcf8574_drv_cfg *drv_cfg = dev->config;
-	struct pcf8574_drv_data *drv_data = dev->data;
+	const struct pcf857x_drv_cfg *drv_cfg = dev->config;
+	struct pcf857x_drv_data *drv_data = dev->data;
 	int rc = 0;
-	uint8_t rx_buf;
+	uint16_t rx_buf;
+	uint8_t rx_buf_helper[2];
+	uint16_t num_gpios = drv_cfg->common.port_pin_mask;
 
-	rc = i2c_read_dt(&drv_cfg->i2c, &rx_buf, sizeof(rx_buf));
-	if (rc != 0) {
-		LOG_ERR("%s: failed to read from device: %d", dev->name, rc);
-		return -EIO;
+	if (num_gpios == 0xFF) {
```

Comment:
Have you considered calculating the buffer length at compile time and using that?
Then you could avoid this if/else chain.

Reply:
updated in recent push 0dd04127ae30274178228f26e6c11d3c1f04cee6